### PR TITLE
Fix for escaping markdown characters in property descriptions

### PR DESCRIPTION
**Pull request by jaandrews**
### Prerequisites

- [x] I have added steps to test this contribution in the description below

This fixes
[Problems escaping markdown characters in property description](https://github.com/umbraco/Umbraco-CMS/issues/12100)

### Description

- Tweaked "src/Umbraco.Web.UI.Client/src/views/components/property/umb-property.html" to hide the description until raw text has been processed by the markdown processor.
- Replaced regex logic with a full markdown parser in the "simpleMarkdown" filter. Used the parser from the "[Markdown Editor](https://github.com/umbraco/Umbraco-CMS/blob/v9/contrib/src/Umbraco.Web.UI.Client/src/views/propertyeditors/markdowneditor/markdowneditor.controller.js)" data type, so no new libraries were added.
- Updated the "umbProperty" component to load "lib/markdown/markdown.converter.js" and set a flag to display the description when that script is loaded.
- Removed duplicate filter file "src/Umbraco.Web.UI.Client/src/common/filters/simpleMarkdown.filter.js.js"

To test this, just try various markdown expressions in content type property descriptions and confirm they are parsed correctly. Make sure to test escaped expressions as well (can't provide an example since github doesn't process them properly either, but it's detailed in the issue I linked to above).

### CODE REVIEW

**Comment by abjerner** [comment]
Hi @jaandrews - I haven't tested your code, but I have some points for whoever is going to review/merge your PR:

- I'm not sure what the correct solution is, but maybe it's worth considering handling this in C# instead?

- As property descriptions have supported HTML for years, this could also case problems:

  https://github.com/umbraco/Umbraco-CMS/pull/12345/files#diff-d41f856e84f73e6f895cc322ff2ad69299f192da79ed544d2c4eb7a398d84f6eR15

  Say I entered the HTML manually for a link, I could end up with duplicate `target` and `rel` properties.

- If the `rel="noopener"` is to address external links, it might be better to use `rel="noreferrer"` as we then don't disclose the backoffice location. `noreferrer` also implies `noopener`.

**Comment by nul800sebastiaan** [comment]
Hi @jaandrews, I'll start by saying that I haven't had a chance to take a look at this but I saw the following comment: 

> added Markdown nuget package to the Umbraco.Web.BackOffice project

We already use `MarkdownSharp` so we should take dependencies on yet another Markdown package. 

Additionally I can see the `umbraco.sln` file has changed, I would recommend never checking in a change to a file that you didn't explicitly changed. I know in this case VS did it for you, but you were not actually working on updating the solution file, you were working on property descriptions.

So if you could please revert the `sln` change and attempt to make this work with MarkdownSharp, that would be great!

**Comment by emmagarland** [comment]
Hi @jaandrews 

Firstly thanks again for this PR and apologies for the delay in replying to you. 

I've just rebased this PR to the `contrib` branch, which created some conflicts that I have since resolved and pushed to your branch. 

I've tested the following property content, including some borrowed from the test on [PR 11628](https://github.com/umbraco/Umbraco-CMS/pull/11628):

```
## Heading (h2) #
### Sub-heading (h3) #
#### Sub-sub-heading (h4) #
##### Sub-sub-sub-heading (h5) #
###### Sub-sub-sub-sub-heading (h6) #
*Markdown test italics*
**Markdown test bold**
\*I want stars \*\*****
No \_italic text\_ 
<hr />
1. First item.
1. Second item.
1. Third item.
<br><br>
\\ backslash itself
\` backtick
\* asterisk
\_ underscore
\{ \} curly braces
\[ \] square brackets
\( \) parentheses
\# hash mark
\+ plus sign
\- minus sign (hyphen)
\. dot
\! exclamation mark
<strong>Strong</strong>
<em>Emphasis</em>
<blockquote>Blockquote</blockquote>

Test content from #11628:
Write some text in this field.
**Important!**
This is going to be live on the website
--
If you write something you *don't like*, you have to apologize. 
![Umbraco logo](/umbraco/assets/img/application/logo.png)
[Umbraco HQ](https://www.umbraco.com)
```

The outcome is:
<hr/>

### **Before PR:**

image
image
image

<hr/>

### After PR:

image
image
<hr/>

The only thing that didn't seem to quite work is the ordered list, but that seems to be both before and after the markdown changes. 

I was wondering if you were okay to double check you are happy with my conflict resolutions and then I can progress with merging this PR? 🙂 



Best wishes

Emma

**Comment by jaandrews** [comment]
@emmagarland I don't see any issues with the applied changes, as that's not a file I touched. Must have been something that changed in umbraco since I last updated this fork. As for the ordered list issue, that must be an issue with the markdown javascript package umbraco uses, as all I'm doing is referencing it to process the content (uses "lib/markdown/markdown.converter.js"). Could be that it's thrown off because the numbers in the ordered list are all "1" instead of "1," "2" and "3."

**Comment by emmagarland** [comment]
Hi @jaandrews, thanks for the reply, and for your pull request to escape markdown characters in property descriptions (#12100). 

It has now been successfully merged and aiming for inclusion the 12.1.0 release! 🙌

Thanks, 

Emma

**Comment by ronaldbarendse** [comment]
This PR contains a regression with rendering the 'Read more' link that allows expanding the description (see PR https://github.com/umbraco/Umbraco-CMS/pull/11628) ⚠️

Besides that, removing the `simpleMarkdown` Angular filter is a breaking change, as this could be used in other places (custom implementations, packages, label templates, etc.).

I also think we shouldn't support all Markdown syntax in the property descriptions, since the formatting can break the layout and cause all sorts of UI and/or accessibility issues. We also shouldn't add the `Markdown` NuGet package to the `Umbraco.Cms.Web.BackOffice` project, since these type of dependencies should only exist in the Infrastructure project (maybe @bergmania can give some additional input/guidance on this). On top of this, we should probably get rid of this dependency, since it was [archived 3 years ago](https://github.com/hey-red/Markdown)...

@nul800sebastiaan Maybe your comment on issue https://github.com/umbraco/Umbraco-CMS/issues/12100#issuecomment-1059077672 was misinterpreted as 'use an existing Markdown engine', instead of keeping the simple syntax and only fix the escaping issue?

Note that you can already handle the [EditorModel notifications](https://docs.umbraco.com/umbraco-cms/reference/notifications/editormodel-notifications) (like `SendingContentNotification`) yourself if you want to do additional parsing and add support for full Markdown syntax. Plus, if you previously used this notification to set/update the description with basic text or the simple Markdown syntax, that won't work anymore, since you'd need to use/parse HTML instead (which is probably another regression)...

File: src/Umbraco.Web.BackOffice/Security/BackOfficeAntiforgery.cs [hunk 1]
```diff
@@ -33,14 +33,12 @@ public BackOfficeAntiforgery(IOptionsMonitor<GlobalSettings> globalSettings)
         {
             x.HeaderName = Constants.Web.AngularHeadername;
             x.Cookie.Name = Constants.Web.CsrfValidationCookieName;
+            x.Cookie.SecurePolicy = globalSettings.CurrentValue.UseHttps ? CookieSecurePolicy.Always : CookieSecurePolicy.SameAsRequest;
```
Comment (ronaldbarendse) [inline src/Umbraco.Web.BackOffice/Security/BackOfficeAntiforgery.cs:36]: Why is this added? This doesn't seem to be relevant for this PR...

Comment (jaandrews) [reply src/Umbraco.Web.BackOffice/Security/BackOfficeAntiforgery.cs:36]: It's not. That's just to resolve a merge conflict the umbraco team encountered.

As for the changes I made
- Umbraco.Web.Backoffice didn't depend have an existing dependency on Umbraco.Infrastructure, so I added the Markdown library directly to it to avoid messing with dependencies. Probably not a difficult change, but I'm not sure whether there is a reason that dependency didn't already exist.
- It's difficult to manually handle markdown edge cases with regex, which is why I went with the markdown library. Also, any issues parsing the markdown text would be an issue with markdown library and would also impact the built in "Markdown Editor" data type. In my case, I needed to display "**" in the description, which is tricky to handle because it needed to check that there isn't a second "**" after it. Gets a bit messier if there are multiple sets.
- You're right that removing the simpleMarkdown filter is technically a breaking change, though it's not a documented filter and so the risk of its removal breaking something is very low.
- Broken markup was also possible before this change, so making sure that doesn't happen is up to the developer.
- Notifications are certainly another way this could have been handled. Was still learning umbraco 9 at the time I created this PR.

Comment (nul800sebastiaan) [reply src/Umbraco.Web.BackOffice/Security/BackOfficeAntiforgery.cs:36]: If removing the `simpleMarkdown` filter is a breaking change, then we can just re-instate the files even though we're not using them. Other than that, existing markdown in descriptions will just be parsed by the markdown library and shouldn't change.

For the rest I concur with @jaandrews. 

For adding a dependency, maybe it needs to be a call to a library with that dependency, but I also don't know if a dependency on `Umbraco.Infrastructure` from `Umbraco.Web.Backoffice` would be welcomed.

**Comment by nul800sebastiaan** [comment]
Alright, I've given the `INotificationHandler` a go and I must say that the result was pretty impressive. I've taken [this Mardown example file](https://gist.github.com/allysonsilva/85fff14a22bbdf55485be947566cc09e), which is big and dropped it all in the Description of a property. This page had 3 tabs, 15 properties in total, 9 of them with description. 

With the parsing of any non-empty descriptions, the profiler says it took a total of 74ms of the entire page load, less than 0.01% of the total load. I with a less huge Markdown description it would be less than 74ms as well of course. 

I think this is a most excellent solution that people can add to their sites instead of having to rely on built-in parsers that CMS needs to maintain. It's using Markdig (`dotnet add package Markdig`), the latest iteration of the outdated library we currently use. 

In order to preserve backwards compatibility and so that we don't have to do weird gymnastics for dependencies, I'll revert this PR and from now on when people want more powerful Markdown parsing of their property descriptions to use the following `INotificationHandler` example.

Thanks again for the work on this PR all!

```csharp
using Markdig;
using Umbraco.Cms.Core.Composing;
using Umbraco.Cms.Core.Events;
using Umbraco.Cms.Core.Notifications;

namespace ExampleProject;

public class PropertyDescriptionMarkdownParse : INotificationHandler<SendingContentNotification>
{
    public void Handle(SendingContentNotification notification)
    {
        foreach (var variant in notification.Content.Variants)
        {
            foreach (var tab in variant.Tabs)
            {
                if (tab.Properties == null)
                {
                    continue;
                }
                
                foreach (var property in tab.Properties)
                {
                    if (string.IsNullOrWhiteSpace(property.Description) == false)
                    {
                        property.Description = Markdown.ToHtml(property.Description);
                    }
                }
            }
        }
    }
    
    public class RegisterNotificationHandlers : IComposer
    {
        public void Compose(IUmbracoBuilder builder)
        {
            builder.AddNotificationHandler<SendingContentNotification, PropertyDescriptionMarkdownParse>();
        }
    }
}
```

Ps. Yes, that huge markdown file did make the description look funny  (see below), but it didn't mess up the layout of the backoffice at all. 👍 

image

**Comment by jaandrews** [comment]
@nul800sebastiaan  After thinking about this a bit, I don't think relying on the INotificationHandler will be able to resolve the issue I ran into if the "simpleMarkdown" filter is still used. The problem in my case is that I need to display "\*\*" in the description. If I enter "\\\*\\\*" it displays as an empty string. If I try escaping the "\\" too with "\\\\\\\*\\\\\\\*" it displays as "\\\\\\\\\\\\." Using a notification handler will do nothing to fix this because the filter doesn't handle the escaped "\*" properly.

**Comment by nul800sebastiaan** [comment]
I'd suggest in that case, maybe try to find other characters, or encode use  ``` `**` ```?

**Comment by nul800sebastiaan** [comment]
Ah I just looked at the original issue again. Yeah, I don't know, maybe Markdig has this option, or some other library offers the proper escaping.

**Comment by jaandrews** [comment]
I just tested and confirmed that using the ascii code for it does make it render as expected ("&ast"). Bit of a hack, but it works.
